python/lsst/pipe/base/graph/graph.py

```diff
@@ -0,0 +1,779 @@
+        # not a node we actually want so remove it here. This is a consequence
+        # of using add_edges from and is an implementation detail.
+        if None in self._connectedQuanta.nodes():
+            self._connectedQuanta.remove_node(None)
+
+    @property
+    def taskGraph(self) -> nx.DiGraph:
+        """Returns a graph representing the relations between the tasks inside
+        the quantum graph.
+
+        Returns
+        -------
+        taskGraph : `networkx.Digraph`
+            Internal datastructure that holds relations of `TaskDefs`s
+        """
+        return self._taskGraph
+
+    @property
+    def graph(self) -> nx.DiGraph:
+        """Returns a graph representing the relations between all the
+        `QuantumNode`s. Larely it should be preferred to iterate over, and use
+        methods of this class, but sometimes direct access to the networkx
+        object may be helpful
+
+        Returns
+        -------
+        graph : `networkx.Digraph`
+            Internal datastructure that holds relations of `QuantumNode`s
+        """
+        return self._connectedQuanta
+
+    def taskDefForNode(self, node: QuantumNode) -> TaskDef:
+        """Retrives the TaskDef associated with a `QuantumNode`.
+
+        Parameters
+        ----------
+        node : `QuantumNode`
+            Node for which the associated `TaskDef` is to be found
+
+        Returns
+        -------
+        `TaskDef`
+            The requested `TaskDef`
+        """
+        return self._quantumToTask[node.nodeNumber]
+
+    def getQuantumNodeByNodeNumber(self, nodeNumber: int) -> QuantumNode:
+        """Lookup a `QuantumNode` from a number associated with the node.
+
+        Parameters
+        ----------
+        nodeNumber : `int`
+            The number associated with a node
+
+        Returns
+        -------
+        `QuantumNode`
+            The node corresponding with input number
+
+        Raises
+        ------
+        `IndexError`
+            If the requested nodeNumber is greater than the number of nodes in
+            the graph.
+        """
+        return self._nodeList[nodeNumber]
+
+    @property
+    def inputQuanta(self) -> List[QuantumNode]:
+        """Makes a `list` of all `QuantumNode`s that are 'input' nodes to the
+        graph, meaning those nodes to not depend on any other nodes in the
+        graph.
+
+        Returns
+        -------
+        inputNodes : list of `QuantumNode`
+        """
+        return [q for q, n in self._connectedQuanta.in_degree if n == 0]
+
+    @property
+    def outputQuanta(self) -> List[QuantumNode]:
+        """Makes a `list` of all `QuantumNode`s that are 'output' nodes to the
+        graph, meaning those nodes have no nodes that depend them in the graph.
+
+        Returns
+        -------
+        outputNodes : list of `QuantumNode`
+        """
+        return [q for q, n in self._connectedQuanta.out_degree if n == 0]
```

Comment:
If this needs to return a container because the convenience is more important than guarding against it being a very large list, I'd recommend a tuple rather than a `list`, because it's a property and it's all too easy to think e.g. `graph.outputQuanta.append(quantum)` is doing something useful if it doesn't raise.  It's not always worthwhile to guard against that, but doing it here is trivial.

But it's also worth asking whether this should just be a generator or return a generator expression.

Same comments apply (sometimes about `set` vs. `frozenset` vs. a lazy iterator) in many other accessors below.

Reply:
I looked and changed this were it seemed appropriate